doc/reference/trd-appid.md

```diff
@@ -0,0 +1,795 @@
+so that an `Option<ShortID>` can be 32 bits in size, with 0 reserved
+for `None`.
+
+```rust
+#[derive(Clone, Copy, Eq)]
+struct ShortID {
+    id: core::num::NonZeroU32
+}
+
+pub trait Compress {
+    fn to_short_id(process: &dyn Process) -> Option<ShortID>;
+}
+```
+
+The `to_short_id` method returns an `Option` so that it has a clear
+default action if the Process Checker does not wish to assign it any
+meaningful identifier that might allow additional access.  A return
+value of `None` semantically means that the process's Application
+Credentials or other state do not map to any known security group or
+set of privileges. A`Some` result means the credentials map to a known
+security group or set of privileges.
+
+Generally, the Process Checker that implements `AppCredentialsChecker`
+and `AppIdentification` also implements `Compress`. This allows it to
+share copies of public keys or other credentials that it uses to make
+decisions, reducing flash space dedicated to these constants. Doing so
+also makes it less likely that the two are inconsistent.
+ 
+`ShortID` values MUST be locally unique among running processes.  The
+mapping between Global Application Identifiers and `ShortID` values
+MUST be deterministic.  Kernels SHOULD implement `Compress` in a
+manner that minimizes the chance that two different Application
+Identifiers compress to the same Short ID (e.g., taking the low-order
+bits of a strong cryptographic hash function, or using a known,
+deterministic mapping).
+
+Short IDs are locally unique for three reasons. First, it simplifies
+process management and naming: a particular application identifier
+uniquely identifies a running process. Second, it ensures that
+resources bound to an application identifier (such as non-volatile
+storage) do not have to handle concurrent accesses from multiple
+processes. Finally, generally one does not want two copies of the same
+Application running: they can create conflicting responses and
+behaviors.
+
+For example, suppose there is a system that wants to grant extra
+permissions to Tock binaries with a `TbfFooterV2Credentials` of
+`Rsa4096Key` with a certain public key. The public key is the Global
+Application Identifier of the process binary. Note this means only one
+process signed with that key can run at any time.
+
+The Process Checker implementing `AppCredentialsChecker` and
+`Compress` stores a copy of this key. It returns `Accept` to calls to
+`check_credentials` with valid `TbfFooterV2Credentials` using this
+key. Calls to `Compress` return `None` for all credentials except a
+`Rsa4096Key` with this key, for which it returns `ShortID {id:
+1}`. The Process Checker also has a method `privileged_id`, which
+returns `ShortID {id: 1}`.
+
+Kernel modules which want to give these processes extra permissions
+can check whether the `ShortID` associated with a process matches the
+`ShortID` returned from `privileged_id`. Alternatively, when they are
```

Comment:
The current mechanisms allow headers in a TBF object to specify permissions; this pushes the definition of privileges to processes and the tools that generate credentials for them. I.e., something like `elf2tab`, when it signs a Userspace Binary, needs to check that the permissions in its headers are correct and is responsible for security policies. The mechanism here is so that the kernel can impose security policies. E.g., only Userspace Binaries signed by Western Digital can access storage, as opposed to any process signed by a key that Western Digital's kernel accepts.

Reply:
> The current mechanisms allow headers in a TBF object to specify permissions; this pushes the definition of privileges to processes and the tools that generate credentials for them. I.e., something like `elf2tab`, when it signs a Userspace Binary, needs to check that the permissions in its headers are correct and is responsible for security policies.

Agreed!


As discussed on https://groups.google.com/g/tock-dev/c/aduN7fHWXdI/m/P0H37in4BgAJ we also decided to allow the kernel to enforce these as well. So the kernel can also be responsible for enforcing security policies.

> The mechanism here is so that the kernel can impose security policies

I must have missed the implementation. Can you point it out?